src/app/CommandSender.cpp

```diff
@@ -657,5 +671,30 @@ void CommandSender::MoveToState(const State aTargetState)
     ChipLogDetail(DataManagement, "ICR moving to [%10.10s]", GetStateStr());
 }
 
+void CommandSender::RollbackData::Checkpoint(CommandSender & aCommandSender)
+{
+    VerifyOrReturn(aCommandSender.mBufferAllocated);
+    VerifyOrReturn(aCommandSender.mState == State::Idle || aCommandSender.mState == State::AddedCommand);
+    VerifyOrReturn(aCommandSender.mInvokeRequestBuilder.GetInvokeRequests().GetError() == CHIP_NO_ERROR);
+    VerifyOrReturn(aCommandSender.mInvokeRequestBuilder.GetError() == CHIP_NO_ERROR);
+    aCommandSender.mInvokeRequestBuilder.Checkpoint(mBackupWriter);
+    mBackupState     = aCommandSender.mState;
+    mRollbackIsValid = true;
+}
+
+CHIP_ERROR CommandSender::RollbackData::Rollback(CommandSender & aCommandSender)
+{
+    VerifyOrReturnError(mRollbackIsValid, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(aCommandSender.mState == State::AddingCommand, CHIP_ERROR_INCORRECT_STATE);
+    ChipLogDetail(DataManagement, "Rolling back response");
+    // TODO(#30453): Rollback of mInvokeRequestBuilder should handle resetting
+    // InvokeResponses.
```

Comment:
What does InvokeResponses have to do with anything here?

Reply:
Updated to InvokeRequest